### PR TITLE
Support for ATSAM arm processors

### PR DESCRIPTION
**Pull request by CrustyAuklet**
Add python scripts to download and parse `.atdf' files for Microchip SAM devices.
Needed by modm-io/modm#189

### CODE REVIEW

File: tools/generator/dfg/sam/sam_identifier.py
Comment (salkinium) [inline tools/generator/dfg/sam/sam_identifier.py:20]: For reference, this is how the datasheet splits it up:
58F9B61D-3D0D-4A77-B549-165122E49CFF

I’d split it a little differently:
```
    sam      d        21    e     15      a     -    m
{platform}{family}{series}{pin}{flash}{revision}-{package}
```
You also need the `{package}`, since they bond the same silicon die differently for the packages, and then some pins may not be available in a different package.

The temperature grade and carrier likely don't have an impact on the device structure, so I'd leave them out for now (we can simply add them later if they turn out important).

File: tools/generator/dfg/sam/avr_mcu.py
Comment (salkinium) [inline tools/generator/dfg/sam/avr_mcu.py]: This file is not needed for Cortex-M based devices, this is only to map the device to the Avrdude name.

File: tools/generator/dfg/sam/sam_identifier.py
Comment (salkinium) [inline tools/generator/dfg/sam/sam_identifier.py]: This won't match, since the above `string = string.lower()`.

File: tools/generator/dfg/sam/sam_identifier.py
Comment (salkinium) [inline tools/generator/dfg/sam/sam_identifier.py]: You also need to set the format string to `{platform}{family}{series}{pin}{flash}{revision}-{package}`, otherwise the identifier cannot format itself. This string is also used to reverse match the input string without having to define a parser for it.

See also: https://blog.salkinium.com/modm-devices/#device-identifiers

File: tools/generator/dfg/sam/avr_registers.py
Comment (salkinium) [inline tools/generator/dfg/sam/avr_registers.py]: You'll probably not need this for SAM.

File: tools/generator/dfg/sam/avr_io.py
Comment (salkinium) [inline tools/generator/dfg/sam/avr_io.py]: Not needed for SAM.

File: tools/generator/dfg/sam/avr_device_tree.py
Comment (salkinium) [inline tools/generator/dfg/sam/avr_device_tree.py]: So the way this works is you get data from the vendor XML using XPath queries using etree Python interface.

You dump this into a dictionary of properties (=`p`) and this is like a vendor-specific internal interface that you can change as you wish. You can use pass this dictionary to other functions and modify it, BUT it only exists for one unique device.

Unfortunately there aren't any helper tools for *exploring* the vendor data set, you'll just have to look at a lot of XML. See also: https://blog.salkinium.com/modm-devices/#peeking-into-stm32cubemx

File: tools/generator/dfg/sam/avr_device_tree.py
Comment (salkinium) [inline tools/generator/dfg/sam/avr_device_tree.py]: This is the function that takes this vendor specific internal data and formats it into our device file format, which is tree based.
This tree has a few intrinsic properties:
- it is **sortable** so that the output is reproducible (very important for checking if something changed via git),
- it is **mergeable**, so that we can reduce the data duplication and overlay multiple data trees of very similiar devices.

This allows us to reduce the number of unique data tree by ~20 fold, which makes it significantly easier to perform manual spot checks to verify the data pipeline.
See also: https://blog.salkinium.com/modm-devices/#encoding-commonality for explanation and examples.

File: tools/generator/dfg/sam/sam_groups.py
Comment (salkinium) [inline tools/generator/dfg/sam/sam_groups.py:18]: These groups are passed to the merger, which will perform the actual device tree merging process. This is a bit try and error to see what looks good, we don't want too much noise in the resulting device files.
See also: https://blog.salkinium.com/modm-devices/#data-pipeline.

**Comment by salkinium** [comment]
I recommend starting with something simple like memory origin, size and type and build your pipeline around that. Then expand that with more data over time.

File: tools/generator/dfg/sam/sam_identifier.py
Comment (CrustyAuklet) [reply tools/generator/dfg/sam/sam_identifier.py:20]: That looks like the old Atmel Datasheet, the new one is [here](http://ww1.microchip.com/downloads/en/DeviceDoc/SAMD21-Family-DataSheet-DS40001882D.pdf)

As far as I can tell the package does not make a difference, with one exception, the WLCSP package. The `.atdf` files reflect this since they either have no package type in the file name or there is a `U` appended to the end.

image

Comment (salkinium) [reply tools/generator/dfg/sam/sam_identifier.py:20]: So then it would just be like this?
```
   atsam     d        21    e     15      a
{platform}{family}{series}{pin}{flash}{revision}
```

Comment (CrustyAuklet) [reply tools/generator/dfg/sam/sam_identifier.py:20]: I was assuming that the name was coming from the `atdf` filename, which I think was a bad assumption. the XML _inside_ the `atdf` file contains full part numbers. so:
```
    sam       d      21     e    15       a     -    m
{platform}{family}{series}{pin}{flash}{revision}-{package}
```
then group according to U and NON-U in `sam_groups.py`

Also, would you prefer for the capture names to match the datasheet or the other python files?
example: change series->name, flash->size like in the stm32 parser

Comment (salkinium) [reply tools/generator/dfg/sam/sam_identifier.py:20]: No, namings should be as close as possible/reasonable to the datasheets. Apart from `{platform}` there is no cross-vendor naming scheme. It's then up to the HAL implementation to rename/reformat the names into a consistent form.

File: tools/generator/dfg/sam/sam_groups.py
Comment (CrustyAuklet) [reply tools/generator/dfg/sam/sam_groups.py:18]: right now it is not merging, but I am getting XML files.
what are the groupings based on?
It seems like SAM might need a `sam_peripherals.py` similar to STM32. From my experience with D21, L21, and D51 the SERCOM driver is compatible for all three except the clocks get set up slightly differently on L21/D51 and L21/D51 have a low power mode. Right now that maps to a single preprocessor #if  in the init function and a low power mode function that does nothing for D21. that might be two different SERCOM implementations?

Comment (salkinium) [reply tools/generator/dfg/sam/sam_groups.py:18]: Maybe try lower-case `d` for family?
The groupings are established manually (the resulting merged tree “looks structured” -> good group), I thought about an automatic heuristic approach, but trial and error approach is so much easier.

Yes, you will want to add a `sam_peripherals.py`. Currently it is the convention to prefix the driver type with the `{platform}-type` or just `{platform}` if its unique, so that the lbuild modules do not need to check both the identifier and the driver type.

Comment (salkinium) [reply tools/generator/dfg/sam/sam_groups.py:18]: Regarding the SERCOM, if the difference is so minor, make them the same type, but maybe add `feature` elements, which can distinguish smaller differences between peripherals. However, we usually only find these differences during the HAL implementation, and then encode this in the HAL until it becomes messy and then move it to the device files. So don’t sweat the details for now.

**Comment by CrustyAuklet** [comment]
I am having one issue with merging. I get the following in a device file for the saml21:
```xml
<gpio device-pin="j" port="B" pin="13">
    <signal driver="eic" name="extint13"/>
    <signal driver="gclk" name="io7"/>
    <signal driver="ptc" name="x13"/>
    <signal device-flash="16|17|18" device-variant="b" device-package="m" driver="ptc" name="y14"/>
    <signal device-flash="17|18" device-variant="b" device-package="u" driver="ptc" name="y14"/>
    <signal driver="sercom" instance="4" name="pad1"/>
    <signal driver="tc" instance="0" name="wo1"/>
    <signal driver="tc" instance="c0" name="wo7"/>
</gpio>
```
The issue is because there is no flash size `16` in the package `u`. 

seems to be the same issue in a different location:
```xml
      <vector position="10" name="SERCOM2"/>
      <vector position="11" name="SERCOM3"/>
      <vector device-pin="e|g|j" device-variant="b" position="12" name="SERCOM4"/>
      <vector device-pin="g|j" device-variant="a" position="12" name="SERCOM4"/>
      <vector device-pin="e|g|j" device-variant="b" position="13" name="SERCOM5"/>
      <vector device-pin="g|j" device-variant="a" position="13" name="SERCOM5"/>
```
pin count e just doesn't exist in variant a.

Is there some better way to group these?

**Comment by salkinium** [comment]
> I am having one issue with merging.

This is still fine as long as this isn't everywhere. The merger will not create ambiguos trees (which is forcibly checked via the DeepDiff later in the script), so while this isn't pretty, it's not a technical issue either. The trees will be correctly split up again.

That being said, these filters do seem a bit arbitrary, I mean, why would the PTC Y14 signal depend on the flash size or package? Is this an issue in the raw data? For AVRs I had to clean up [the raw data too](https://github.com/modm-io/modm-devices/blob/develop/tools/generator/raw-data-extractor/patches/avr.patch). Maybe also check for some syntax errors in the XML, the AVRs had quite a lot of that.

> Is there some better way to group these?

You can split up the devices files a little more if you want, by for example splitting out the package `u` for L21 devices. The device merger will pick the first match of the group list, so you can put more specific matches like this in the beginning.

But there should be a balance between duplicated trees (in separate files) and ugly filters (in the same file), so that it's easier for us to see such issues. So if those are the only two issues, then just leave them in the same group for now, as I suspect an issue in the raw data.

**Comment by CrustyAuklet** [comment]
> That being said, these filters do seem a bit arbitrary, I mean, why would the PTC Y14 signal depend on the flash size or package? Is this an issue in the raw data? For AVRs I had to clean up [the raw data too](https://github.com/modm-io/modm-devices/blob/develop/tools/generator/raw-data-extractor/patches/avr.patch). Maybe also check for some syntax errors in the XML, the AVRs had quite a lot of that.

On further inspection it looks like the source data actually doesn't list sercom4/5 for the smallest pin count variant A devices. This isn't documented in the datasheet but variant A means engineering sample so... 🤷‍♂️. It could be an omission but it doesn't look like a typo.

the other issue `device-flash="16|17|18"` vs `device-flash="17|18"` just seems like the merger is upset that size 16 doesn't exist in package u. It only happens 3 times though so I won't worry about it.

These devices merge pretty nicely otherwise. I was doing some datasheet and XML comparisons this morning and it seems like the only difference between chips in the same family (ignoring small changes between revisions) is pinout and the number of sercoms/timers.

**Comment by salkinium** [comment]
> These devices merge pretty nicely otherwise.

Ok, great! So the next steps for this repo:

1. Commit the generated device files.
2. Make this run in the CI (requires to un-WIP this PR cos Travis is weird).
3. I won't merge this just yet, because there will probably be more changes once you start the HAL implementation in modm.

Next steps for your modm fork:
1. Add `ext/microchip/cmsis-header-atsam` submodule, update the `ext/modm-devices` submodule to point to yours temporarily.
2. Start your implementation (I will comment on that in the modm repo), open PR, pass the CI, etc.
3. Fix issues in the modm-devices data that appear during HAL implementation.
4. Once we're happy with both repos, first merge this PR, then merge the modm PR. You're done!

So there is some iteration involved to validate this data for the usage in the HAL, unfortunately it's spread across three repos, so its a bit more work, but that is the cost of modularization.

At some point I'd like to move your `cmsis-header-atsam` to the `modm-io` org, with you still as owner. To have all the parts in one place makes it easier to maintain.

File: devices/sam/samd21.xml
Comment (salkinium) [inline devices/sam/samd21.xml]: Filter these out, they are Cortex-M specific and [added by the `modm:platform:cortex-m` module](https://github.com/modm-io/modm/blob/develop/src/modm/platform/core/cortex/module.lb#L36-L55) (for now).

File: devices/sam/samd21.xml
Comment (salkinium) [inline devices/sam/samd21.xml]: Make this `cortex-m0+` to [align with STM32](https://github.com/modm-io/modm-devices/blob/develop/devices/stm32/stm32l0-51_52_53_62_63.xml#L41).

File: devices/sam/samd21.xml
Comment (salkinium) [inline devices/sam/samd21.xml:87]: Wonky.

File: devices/sam/samd21.xml
Comment (salkinium) [inline devices/sam/samd21.xml]: Wonky. What is `hmatrixb` anyway? I can't find it in the datasheet.

File: devices/sam/samd21.xml
Comment (salkinium) [inline devices/sam/samd21.xml:5]: What happened to the package type? You can use an empty string ([like the AVRs](https://github.com/modm-io/modm-devices/blob/develop/devices/avr/atmega-48_88_168_328-a_n_p_pa.xml#L10)), but it seems like there is some default package type (maybe A = TQFP) that isn't explicitly declared.

File: devices/sam/samd51.xml
Comment (salkinium) [inline devices/sam/samd51.xml]: `cortex-m4f`

File: devices/sam/samd51.xml
Comment (salkinium) [inline devices/sam/samd51.xml]: Remove Cortex-M specific handlers

File: devices/sam/samd51.xml
Comment (salkinium) [inline devices/sam/samd51.xml]: wonky

File: tools/generator/dfg/sam/sam_device_tree.py
Comment (salkinium) [inline tools/generator/dfg/sam/sam_device_tree.py]: The `topLevelOrder` can be removed, same with `attribute-mcu`. They are left over from the AVRs.

File: devices/sam/samd21.xml
Comment (salkinium) [inline devices/sam/samd21.xml]: No SRAM?

Comment (salkinium) [reply devices/sam/samd21.xml]: What about start addresses? You'll need that later for the [linkerscript generator in `modm:platform:cortex-m`](https://github.com/modm-io/modm/blob/develop/src/modm/platform/core/cortex/linker.macros#L18-L25). They differ for device variant, due to RWW.

**Review by salkinium — COMMENTED** [review]
I've added a few comments mostly about missing and consistent data, but otherwise this looks *really good* already!

**Comment by salkinium** [comment]
> make: execvp: ./sam_generator.py: Permission denied

Make sure the file is marked as executable (`chmod a+x sam_generator.py`).

File: devices/sam/samd21.xml
Comment (CrustyAuklet) [reply devices/sam/samd21.xml:5]: I was having a lot of issues with package. it is only specified if it matters and the only thing it changes is the pinout. For the three chips I am working with now only the `u` package is unique. There is a list of compatible variants at the beginning of each `atsd` file, but then there is a lot of duplication. So I just made the variant include the package, so the identifier now has a 1 or 2 character variant. This makes it match the way microchip names their files as well.

Initally I had all packages that didn't specify the package default to `m`. but that might be confusing for someone looking for the `a` package. I am also not sure it will scale. I don't use other SAM chips but there might be other chips where m is the weird one and u is the default.

File: devices/sam/samd21.xml
Comment (CrustyAuklet) [reply devices/sam/samd21.xml]: it appears to be a part of the SDHC module? in the `atsd` file it does't have any signals only attributes so I should probably filter it out.

File: devices/sam/samd21.xml
Comment (salkinium) [reply devices/sam/samd21.xml:5]: Hm, I'd still make it a separate *key* (with an empty string for no package like `package="|u"`). The reason is that then you don't drag this hack into the Jinja templates as well, because the identifier is used to split up the `modm:target` string into it's components.
```
%% if target.variant in ['a', 'au', 'b', 'bu']
    enum class VariantABonly {};
%% endif
```
Instead of:
```
%% if target.variant in ['a', 'b']
    enum class VariantABonly {};
%% endif
```

But lets leave this for some later iteration, since this may not even be that important.

File: devices/sam/samd21.xml
Comment (CrustyAuklet) [reply devices/sam/samd21.xml:87]: `l` variants are "optimized for analog and pwm" so they have two AC modules, it should read:
```
<driver name="ac" type="sam">
<driver Device-variant="l" name="ac1" type="sam">
```

Comment (salkinium) [reply devices/sam/samd21.xml:87]: Oh, I see. And the datasheet refers to them as `AC` and `AC1` for L variants, so this is actually correct.

File: devices/sam/samd21.xml
Comment (CrustyAuklet) [reply devices/sam/samd21.xml]: everything has a different name... I'm sure this will be fun when I add more variants. done.

File: tools/generator/dfg/sam/sam_device_tree.py
Comment (salkinium) [inline tools/generator/dfg/sam/sam_device_tree.py]:
```suggestion
            access = memory_segment.get('rw').lower()
```

File: tools/generator/dfg/sam/sam_device_tree.py
Comment (salkinium) [inline tools/generator/dfg/sam/sam_device_tree.py]:
```suggestion
        core = device_file.query("//device")[0].get('architecture').lower().replace("plus", "+")
```

**Review by salkinium — DISMISSED** [review]
Excellent! Onwards to the HAL!!

**Comment by salkinium** [comment]
I'm fairly confident in your work and I want to merge it before it goes stale.
There are some changes to modm-devices on the horizon, and I want to have more data and more devices to play with to catch the corner cases. I didn't want to continuously force push onto *your* develop branch, therefore I've opened up my own fork.

**Comment by salkinium** [comment]
Merged with minimal changes via #31.

**Pull request closed without merging**
